### PR TITLE
feat: implement proof verification

### PR DESCRIPTION
**Pull request by nmlinaric**
## Description
Implement proof verification logic for validate on destination that a deposit happened on source chain.

## Related Issue Or Context
Closes: #5 

## How Has This Been Tested? Testing details.
unit test, e2e test, manual tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [x] I have updated the documentation locally and in chainbridge-docs.
- [x] I have added tests to cover my changes.
- [x] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

File: src/contracts/Executor.sol [hunk 1]
```diff
@@ -18,95 +22,135 @@ contract Executor is Context {
...
         _;
     }
 
-    constructor(address bridge) {
+    function _onlyAllowed(bytes4 sig, address sender) private view {
+        if (!_accessControl.hasAccess(sig, sender)) revert AccessNotAllowed(sender, sig);
+    }
+
+    constructor(
+        uint8 originDomainID,
```
Comment (mpetrun5) [inline src/contracts/Executor.sol:69]: This should either be an array or you should remove it and just use the function.

File: src/contracts/BlockStorage.sol [hunk 2]
```diff
@@ -0,0 +1,14 @@
+pragma solidity 0.8.11;
+
+
+contract BlockStorage {
```
Comment (mpetrun5) [inline src/contracts/BlockStorage.sol:4]: This is extra. Create an interface for the `SpectreProxy` contract and all other state root storage contracts.
`IStateRootStorage`

File: src/contracts/Executor.sol [hunk 3]
```diff
@@ -118,4 +162,42 @@ contract Executor is Context {
...
+        - data Data originally provided when deposit was made.
+        @return Boolean value depending if deposit nonce has already been used or not.
+     */
+    function verify(Proposal memory proposal, bytes32 storageRoot)
+        internal
+        returns (bool)
+    {
+        bytes32 transferHash;
+
+        (bytes[] memory storageProof) = proposal.storageProof;
```
Comment (mpetrun5) [inline src/contracts/Executor.sol:181]: No need to initialize this.

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: src/contracts/Executor.sol [hunk 4]
```diff
@@ -18,95 +22,132 @@ contract Executor is Context {
...
         for (uint256 i = 0; i < proposals.length; i++) {
             if (isProposalExecuted(proposals[i].originDomainID, proposals[i].depositNonce)) {
                 continue;
             }
+            bytes32 stateRoot;
+            bytes32 storageRoot;
+            address routerAddress = _originDomainIDToRouter[proposals[i].originDomainID];
 
+            IStateRootStorage blockHeaderStorage = IStateRootStorage(_securityModels[proposals[i].securityModel]);
+            stateRoot = IStateRootStorage(blockHeaderStorage).getStateRoots(proposals[i].originDomainID, slot);
```
Comment (lastperson) [inline src/contracts/Executor.sol:118]:
```suggestion
            stateRoot = blockHeaderStorage.getStateRoots(proposals[i].originDomainID, slot);
```

File: src/contracts/Executor.sol [hunk 5]
```diff
@@ -18,95 +22,132 @@ contract Executor is Context {
     using ECDSA for bytes32;
 
     IBridge public immutable _bridge;
+    uint8 public immutable _domainID;
+    uint256 public _slotIndex;
```
Comment (lastperson) [inline src/contracts/Executor.sol:26]:
```suggestion
    uint256 public immutable _slotIndex;
```

**Review by lastperson — CHANGES_REQUESTED** [review]

File: src/contracts/Executor.sol [hunk 6]
```diff
@@ -18,95 +22,130 @@ contract Executor is Context {
     using ECDSA for bytes32;
 
     IBridge public immutable _bridge;
+    uint8 public immutable _domainID;
+    uint8 public constant SLOT_INDEX = 2;
+    IAccessControlSegregator public _accessControl;
 
+    // securityModel => block header storage address
```
Comment (mpetrun5) [inline src/contracts/Executor.sol:29]:
```suggestion
    // securityModel => state root storage contract address
```

File: src/contracts/Executor.sol [hunk 7]
```diff
@@ -18,95 +22,130 @@ contract Executor is Context {
...
         bytes32 resourceID;
         bytes data;
+        bytes[] storageProof;
     }
 
-    event ProposalExecution(uint8 originDomainID, uint64 depositNonce, bytes32 dataHash, bytes handlerResponse);
-    event FailedHandlerExecution(bytes lowLevelData, uint8 originDomainID, uint64 depositNonce);
+    event ProposalExecution(uint8 originDomainID, uint64 depositNonce);
+    event FeeRouterChanged(uint8 originDomainID, address newRouter);
+    event SlotIndexChanged(uint256 slotIndex);
```
Comment (mpetrun5) [inline src/contracts/Executor.sol:47]:
```suggestion
```

File: src/contracts/Executor.sol [hunk 8]
```diff
@@ -18,95 +22,130 @@ contract Executor is Context {
...
 
         for (uint256 i = 0; i < proposals.length; i++) {
             if (isProposalExecuted(proposals[i].originDomainID, proposals[i].depositNonce)) {
                 continue;
             }
+            bytes32 stateRoot;
+            bytes32 storageRoot;
+            address routerAddress = _originDomainIDToRouter[proposals[i].originDomainID];
 
+            IStateRootStorage blockHeaderStorage = IStateRootStorage(_securityModels[proposals[i].securityModel]);
```
Comment (mpetrun5) [inline src/contracts/Executor.sol:115]:
```suggestion
            IStateRootStorage stateRootStorage = IStateRootStorage(_securityModels[proposals[i].securityModel]);
```

File: src/contracts/Executor.sol [hunk 9]
```diff
@@ -18,95 +22,130 @@ contract Executor is Context {
...
-                    dataHash,
-                    handlerResponse
-                );
-            } catch (bytes memory lowLevelData) {
-                emit FailedHandlerExecution(lowLevelData, proposals[i].originDomainID, proposals[i].depositNonce);
-                usedNonces[proposals[i].originDomainID][proposals[i].depositNonce / 256] &= ~(1 <<
-                    (proposals[i].depositNonce % 256));
-                continue;
-            }
+            depositHandler.executeProposal(proposals[i].resourceID, proposals[i].data);
```
Comment (mpetrun5) [inline src/contracts/Executor.sol:125]: Why don't we keep the try catch logic?
Don't we still need the usedNonces revert when a proposal execution fails.

File: src/contracts/Router.sol [hunk 10]
```diff
@@ -24,6 +24,10 @@ contract Router is Context {
     IAccessControlSegregator public _accessControl;
     uint8 public immutable _domainID;
 
+    // securityModel => block header storage address
+    mapping(uint8 => address) public _securityModels;
```
Comment (mpetrun5) [inline src/contracts/Router.sol:28]: Why do we need this here?

File: src/contracts/Executor.sol (hunk 9 above)
Comment (mpetrun5) [reply src/contracts/Executor.sol:125]: Why have we removed executeProposal bytes that were returned?

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: src/contracts/Executor.sol (hunk 9 above)
Comment (nmlinaric) [reply src/contracts/Executor.sol:125]: A misunderstanding, understood we're removing it :see_no_evil:

**Review by mpetrun5 — APPROVED** [review]

**Review by lastperson — APPROVED** [review]

**Pull request merged**